pep-0xxx.rst

```diff
@@ -0,0 +1,571 @@
+            """
+            pass
+
+        @abstractmethod
+        def cipher(self) -> Optional[Cipher]:
+            """
+            Returns the Cipher entry for the cipher that has been negotiated on
+            the connection. If no connection has been negotiated, returns
+            ``None``.
+            """
+            pass
+
+        @abstractmethod
+        def negotiated_protocol(self) -> Optional[str]:
+            """
+            Returns the protocol that was selected during the TLS handshake.
+            This selection may have been made using ALPN, NPN, or some future
+            negotiation mechanism.
+
+            If ``Context.set_inner_protocols()`` was not called, if the other
+            party does not support protocol negotiation, if this socket does
+            not support any of the peer's proposed protocols, or if the
+            handshake has not happened yet, ``None`` is returned.
+            """
+            pass
+
+        @property
+        @abstractmethod
+        def context(self) -> Context:
+            """
+            The ``Context`` object this socket is tied to.
+            """
+            pass
+
+        @context.setter
+        @abstractmethod
+        def context(self, value: Context) -> None:
+            """
+            Set the value of the ``Context`` object this socket is tied to.
+            This operation (changing the context) may not always be supported.
+            """
+            pass
+
+
+Cipher Suites
+~~~~~~~~~~~~~
+
+Todo
+
+Errors
+~~~~~~
+
+This module would define three base classes for use with error handling. Unlike
+the other classes defined here, these classes are not *abstract*, as they have
+no behaviour. They exist simply to signal certain common behaviours. Backends
+should subclass these exceptions in their own packages, but needn't define any
+behaviour for them. These exceptions should *never* be thrown directly, they
+should always be subclassed.
+
+The definitions of the errors are below::
+
+    class TLSError(Exception):
+        """
+        The base exception for all TLS related errors from any backend.
+        Catching this error should be sufficient to catch *all* TLS errors,
+        regardless of what backend is used.
+        """
+        pass
+
+    class WantWriteError(TLSError):
+        """
+        A special signaling exception used only when non-blocking or
+        buffer-only I/O is used. This error signals that the requested
+        operation cannot complete until more data is written to the network,
+        or until the output buffer is drained.
+        """
+        pass
+
+    class WantReadError(TLSError):
+        """
+        A special signaling exception used only when non-blocking or
+        buffer-only I/O is used. This error signals that the requested
+        operation cannot complete until more data is read from the network, or
+        until more data is available in the input buffer.
+        """
+        pass
+
```

Comment:
The main reason I haven't defined a more complete exception hierarchy is because I'm reluctant to constrain what applications can provide. We could certainly provide an abstract validation error though, and then allow implementations to provide more specific subclasses if they are able.

Does that sound sensible?